Added Windows Containerd GMSA Updates (#27131)
kubernetes/website: content/en/docs/tasks/configure-pod-container/configure-gmsa.md

1. The container runtime configures each Windows container with the specified GMSA credential spec so that the container can assume the identity of the GMSA in Active Directory and access services in the domain using that identity.

## Containerd 

> On Windows Server 2019, in order to use GMSA with Containerd, you must be running OS Build 17763.1817 (or later) which can be installed using the patch [KB5000822](https://support.microsoft.com/en-us/topic/march-9-2021-kb5000822-os-build-17763-1817-2eb6197f-e3b1-4f42-ab51-84345e063564). 

Reviewer: nit:
```suggestion
On Windows Server 2019, in order to use GMSA with containerd, you must be running OS Build 17763.1817 (or later) which can be installed using the patch [KB5000822](https://support.microsoft.com/en-us/topic/march-9-2021-kb5000822-os-build-17763-1817-2eb6197f-e3b1-4f42-ab51-84345e063564).
```

> There is also a known issue with containerd when trying to connect to SMB shares from Pods. Once you have configured GMSA, the pod will be unable to connect to the share using the Hostname or FQDN but connecting via IP address works as expected. 

Reviewer: I'd write:
```suggestion
There is also a known issue with containerd that occurs when trying to connect to SMB shares from Pods. Once you have configured GMSA, the pod will be unable to connect to the share using the hostname or FQDN, but connecting to the share using an IP address works as expected. 
```

```PowerShell
ping adserver.ad.local
```

> Resolves the Hostname to IP Address Correctly and will produce an output similar to this:

Reviewer: ```suggestion
correctly resolves the hostname to an IPv4 address. The output is similar to:
```

```

> ```PowerShell

Reviewer: @perithompson please omit the “PowerShell” from this line - it's output, not PowerShell.

Pinging adserver.ad.local [192.168.111.18] with 32 bytes of data:
Reply from 192.168.111.18: bytes=32 time=6ms TTL=124
Reply from 192.168.111.18: bytes=32 time=5ms TTL=124
Reply from 192.168.111.18: bytes=32 time=5ms TTL=124
Reply from 192.168.111.18: bytes=32 time=5ms TTL=124
```

However, when attempting to browse the directory using the hostname

```PowerShell
cd \\adserver.ad.local\test
```

> It will appear that the share doesn't exists:

Reviewer: I'd write:
```suggestion
you see an error that implies the target share doesn't exist:
```

> ```PowerShell

Reviewer: @perithompson please omit the “PowerShell” from this line - it's output, not PowerShell.

cd : Cannot find path '\\adserver.ad.local\test' because it does not exist.
At line:1 char:1
+ cd \\adserver.ad.local\test
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (\\adserver.ad.local\test:String) [Set-Location], ItemNotFoundException
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.SetLocationCommand
```

> While trying to browse to the share using the IP Address instead:

Reviewer: I'd write:
```suggestion
and you notice that the error disappears if you browse to the share using its IPv4 address instead; for example:
```

```PowerShell
cd \\192.168.111.18\test
```

> Changes to the directory as expected and you should see a prompt similar to this:

Reviewer: I'd write:
```suggestion
After you change into a directory within the share, you see a prompt similar to:
```

> ```PowerShell

Reviewer: @perithompson please omit the “PowerShell” from this line - it's a shell prompt, and won't highlight properly as PowerShell code.

Microsoft.PowerShell.Core\FileSystem::\\192.168.111.18\test>
```

> To correct the behaviour you must run the following `reg add "HKLM\SYSTEM\CurrentControlSet\Services\hns\State" /v EnableCompartmentNamespace /t REG_DWORD /d 1` and restart the pods to add the required registry key.

Reviewer 2: Can you clarify that this needs to be run on the node?

Reviewer: What does “restart the pods” mean? Is it in fact Pod deletion followed by the control plane autohealing your workload?

## Troubleshooting

If you are having difficulties getting GMSA to work in your environment, there are a few troubleshooting steps you can take.

First, make sure the credspec has been passed to the Pod.  To do this you will need to `exec` into one of your Pods and check the output of the `nltest.exe /parentdomain` command.  

In the example below the Pod did not get the credspec correctly:

```PowerShell
kubectl exec -it iis-auth-7776966999-n5nzr powershell.exe
```
`nltest.exe /parentdomain` results in the following error:

> ```PowerShell

Reviewer: @perithompson please omit the “PowerShell” from this line - it's not something you type into a shell.

1. The container runtime configures each Windows container with the specified GMSA credential spec so that the container can assume the identity of the GMSA in Active Directory and access services in the domain using that identity.

## Containerd 

> On Windows Server 2019 in order GMSA with Containerd, you must be running OS Build 17763.1817 (or later) which can be installed using the patch [KB5000822](https://support.microsoft.com/en-us/topic/march-9-2021-kb5000822-os-build-17763-1817-2eb6197f-e3b1-4f42-ab51-84345e063564). 

Reviewer 3: in order to use*

Author: Oh man! Yes that should be in order to use!

There is also a known issue with containerd when trying to connect to SMB shares from Pods. Once you have configured GMSA, the pod will be unable to connect to the share using the Hostname or FQDN but connecting via IP address works as expected. 

```shell
Windows PowerShell
Copyright (C) Microsoft Corporation. All rights reserved.
PS C:\> ping adserver.ad.local
Pinging adserver.ad.local [192.168.111.18] with 32 bytes of data:
Reply from 192.168.111.18: bytes=32 time=6ms TTL=124
Reply from 192.168.111.18: bytes=32 time=5ms TTL=124
...
PS C:\> cd \\adserver.ad.local\test
cd : Cannot find path '\\adserver.ad.local\test' because it does not exist.
At line:1 char:1
+ cd \\adserver.ad.local\test
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (\\adserver.ad.local\test:String) [Set-Location], ItemNotFoundException
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.SetLocationCommand
PS C:\> cd \\192.168.111.18\test
PS Microsoft.PowerShell.Core\FileSystem::\\192.168.111.18\test>

> ```

Reviewer: This is not valid Bourne shell.

Please:
- mark the code block as PowerShell
- [omit the prompt](https://kubernetes.io/docs/contribute/style/style-guide/#don-t-include-the-command-prompt)
- [separate commands from output](https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output)

Author: thanks for the feedback @sftim I have corrected those and I also went ahead and fixed the other parts that were not in the correct style as well, please can you take another look?

If you add the `lifecycle` section show above to your Pod spec, the Pod will execute the commands listed to restart the `netlogon` service until the `nltest.exe /query` command exits without error.

## Containerd 

> On Windows Server 2019 in order to use SMB Shares with Containerd, you must be running OS Build 17763.1817 (or later) which can be install using the patch [KB5000822](https://support.microsoft.com/en-us/topic/march-9-2021-kb5000822-os-build-17763-1817-2eb6197f-e3b1-4f42-ab51-84345e063564). 

Reviewer 3: installed*

There is also a known issue with containerd when trying to connect to SMB shares from Pods. In some cases, once you have configured GMSA, the pod will be unable to connect to the share using the Hostname or FQDN but connecting via IP address works as expected. 

```shell
Windows PowerShell
Copyright (C) Microsoft Corporation. All rights reserved.
PS C:\> ping adserver.ad.local
Pinging adserver.ad.local [192.168.111.18] with 32 bytes of data:
Reply from 192.168.111.18: bytes=32 time=6ms TTL=124
Reply from 192.168.111.18: bytes=32 time=5ms TTL=124
...
PS C:\> cd \\adserver.ad.local\test
cd : Cannot find path '\\adserver.ad.local\test' because it does not exist.
At line:1 char:1
+ cd \\adserver.ad.local\test
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (\\adserver.ad.local\test:String) [Set-Location], ItemNotFoundException
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.SetLocationCommand
PS C:\> cd \\192.168.111.18\test
PS Microsoft.PowerShell.Core\FileSystem::\\192.168.111.18\test>
```

> If you encounter a case like above you should run the following `reg add "HKLM\SYSTEM\CurrentControlSet\Services\hns\State" /v EnableCompartmentNamespace /t REG_DWORD /d 1` and restart the pods.

Reviewer 3: I feel like it might make more sense to just state that if you want to access smb shares by fqdn you'll need to turn this on, otherwise the above will happen and not vice versa. It's a pain that we need to turn this on but it was a precaution to not break accidentally break anything :/.

Reviewer 3: I worded that rather oddly. Basically the "known issue" is the one we fixed, and the fix is just gated behind a registry key. So instead of "this won't work" -> "to fix turn on reg key", might make more sense to say "to enable smb access by hostname turn on reg key" -> "otherwise this won't work".

Author: Yeah fair point, when I wrote this last night I was thinking there would be cases when this wasn't needed but on reflection, you're right, this needs to be turned on!

If the DNS and communication test passes, next you will need to check if the Pod has established secure channel communication with the domain.  To do this, again, `exec` into your Pod and run the `nltest.exe /query` command.

```PowerShell
nltest.exe /query
```

Results in the following output:

> ```PowerShell

Reviewer: @perithompson please omit the “PowerShell” from this line - it's not something you type into a shell.

I_NetLogonControl failed: Status = 1722 0x6ba RPC_S_SERVER_UNAVAILABLE
```

This tells us that for some reason, the Pod was unable to logon to the domain using the account specified in the credspec.  You can try to repair the secure channel by running the `nltest.exe /sc_reset:domain.example` command.

```PowerShell
nltest /sc_reset:domain.example
```

If the command is successful you will see and output similar to this:

> ```PowerShell

Reviewer: @perithompson please omit the “PowerShell” from this line - it's not something you type into a shell.

Flags: 30 HAS_IP  HAS_TIMESERV
Trusted DC Name \\dc10.domain.example
Trusted DC Connection Status Status = 0 0x0 NERR_Success
The command completed successfully
```

> If the abovecorrects the error, you can automate the step by adding the following lifecycle hook to your Pod spec.  If it did not correct the error, you will need to examine your credspec again and confirm that it is correct and complete.

Reviewer 2: ```suggestion
If the above corrects the error, you can automate the step by adding the following lifecycle hook to your Pod spec.  If it did not correct the error, you will need to examine your credspec again and confirm that it is correct and complete.
```
